File: problems/0039-combination-sum/python/solution.py
def combinationSum(candidates, target):
    """Combination Sum — https://leetcode.com/problems/combination-sum/"""
    res = []
    combo = []

    def backtrack(start, remain):
        if remain == 0:
            res.append(combo[:])
            return
        if remain < 0:
            return
        for i in range(start, len(candidates)):
            combo.append(candidates[i])
            backtrack(i, remain - candidates[i])  # reuse allowed -> i, not i+1
            combo.pop()

    backtrack(0, target)
    return res

File: problems/0039-combination-sum/typescript/solution.ts
// Combination Sum — https://leetcode.com/problems/combination-sum/
export function combinationSum(candidates: number[], target: number): number[][] {
  const res: number[][] = [];
  const combo: number[] = [];
  const backtrack = (start: number, remain: number): void => {
    if (remain === 0) {
      res.push([...combo]);
      return;
    }
    if (remain < 0) return;
    for (let i = start; i < candidates.length; i++) {
      combo.push(candidates[i]);
      backtrack(i, remain - candidates[i]); // reuse allowed -> i
      combo.pop();
    }
  };
  backtrack(0, target);
  return res;
}
